swarm/storage/resource.go

```diff
@@ -91,186 +114,168 @@ type resource struct {
 // stored using a separate store, and forwarding/syncing protocols carry per-chunk
 // flags to tell whether the chunk can be validated or not; if not it is to be
 // treated as a resource update chunk.
-
-type ResourceValidator interface {
-	isOwner(string) (bool, error)
-	nameHash(string) common.Hash
-}
-
+//
+// TODO: Include modtime in chunk data + signature
 type ResourceHandler struct {
 	ChunkStore
 	validator    ResourceValidator
-	rpcClient    *rpc.Client
+	ethClient    *ethclient.Client
 	resources    map[string]*resource
 	hashLock     sync.Mutex
 	resourceLock sync.RWMutex
 	hasher       SwarmHash
-	privKey      *ecdsa.PrivateKey
-	maxChunkData int64
+	nameHash     nameHashFunc
+	ctx          context.Context
+	cancelFunc   func()
 }
 
 // Create or open resource update chunk store
-func NewResourceHandler(privKey *ecdsa.PrivateKey, datadir string, cloudStore CloudStore, rpcClient *rpc.Client, validator ResourceValidator) (*ResourceHandler, error) {
-	path := filepath.Join(datadir, "resource")
-	dbStore, err := NewDbStore(datadir, nil, singletonSwarmDbCapacity, 0)
+//
+// If validator is nil, signature and access validation will be deactivated
+func NewResourceHandler(datadir string, cloudStore CloudStore, ethClient *ethclient.Client, validator ResourceValidator) (*ResourceHandler, error) {
+
+	hashfunc := MakeHashFunc(SHA3Hash)
+
+	path := filepath.Join(datadir, DbDirName)
+	dbStore, err := NewDbStore(datadir, hashfunc, singletonSwarmDbCapacity, 0)
 	if err != nil {
 		return nil, err
 	}
 	localStore := &LocalStore{
 		memStore: NewMemStore(dbStore, singletonSwarmDbCapacity),
 		DbStore:  dbStore,
 	}
-	hasher := MakeHashFunc("SHA3")
 
+	ctx, cancel := context.WithCancel(context.Background())
 	rh := &ResourceHandler{
-		ChunkStore:   newResourceChunkStore(path, hasher, localStore, cloudStore),
-		rpcClient:    rpcClient,
-		resources:    make(map[string]*resource),
-		hasher:       hasher(),
-		privKey:      privKey,
-		maxChunkData: DefaultBranches * int64(hasher().Size()),
+		ChunkStore: newResourceChunkStore(path, hashfunc, localStore, cloudStore),
+		ethClient:  ethClient,
+		resources:  make(map[string]*resource),
+		hasher:     hashfunc(),
+		validator:  validator,
+		ctx:        ctx,
+		cancelFunc: cancel,
```

Comment:
It is recommended that `context.Context` is passed as the first argument when passing an API boundary, what is the benefit here of setting it as a long lived field value?

Reply:
But it can't be passed before the `ChunkStore` interface, right?

Reply:
My thinking was that a shallow context would enable cancelling any pending GetBlock() rpc backend calls (or other calls) when the `Close` method is called. Is this a problem?

Reply:
It just isn't the how context is intended to be used. Users are supposed to create a context _per_ api call, and that assumption is used in most libraries (e.g. cancelling a context cancels the _current_ API call, not all API calls).